### ISSUE TITLE
CPU race on fetch error

### ISSUE THREAD

**Issue by jdconley** [fluent/fluent-plugin-kafka#129]
We are receiving the following error using the kafka_group input plugin many times per second.

```
2017-05-17 21:28:40 +0000 fluent.error: {"error":"No partitions assigned!","message":"unexpected error during consuming events from kafka. Re-fetch events. error=\"No partitions assigned!\""}
```

This error appears to be raised [by the ruby-kafka library](https://github.com/zendesk/ruby-kafka/blob/51dbfe2261817205415fc2ca8b9336dd115c73df/lib/kafka/consumer.rb#L356) whenever a consumer is not assigned a partition. Some topics only have one partition, but we still use multiple consumers in the group for high availability. This causes some clients to not be assigned partitions unless another one leaves the group.

Can some sort of error backoff timeouts be applied to the input plugin, especially on this particular error? Or can we otherwise more gracefully handle this situation? Maybe a configurable range?

I might raise the issue with ruby-kafka as well, as I'm not sure this is truly an exception case.

**Comment by repeatedly** [comment]
If ruby-kafka will handle these case internally, I want to forward it to ruby-kafka.
But if ruby-kafka always raises an error for this case, plugin should care it.
Maybe, refreshing metadata is needed when receives this error.

**Comment by jdconley** [comment]
I'm not sure how the workflow with kafka should work. But, from this plugin's perspective, it would be helpful if this error did not cause a CPU spike and write hundreds of logs per second. Some sort of backoff on error would be very helpful. Even limiting to one retry per second, or one per five seconds would be enough to not overwhelm the system.

**Comment by jdconley** [comment]
Looks like there is an issue open already in ruby-kafka: https://github.com/zendesk/ruby-kafka/issues/338

**Comment by kenhys** [comment]
It seems that it was fixed in ruby-kafka 0.5.0.
